shared/lib/gcs.py

```diff
@@ -16,81 +16,107 @@
+    """
+  logging.info("Download %s/%s to %s", bucket_name, blob_name, local_path)
   if use_anonymous_client:
     storage_client = storage.Client.create_anonymous_client()
   else:
     storage_client = storage.Client()
-  bucket_object = storage_client.bucket(bucket_name=bucket)
-  blob = bucket_object.get_blob(filename)
-  # Download
-  local_file_path = _get_local_path(bucket, filename)
-  # Create directory to file if it does not exist.
-  parent_dir = Path(local_file_path).parent
-  if not parent_dir.exists():
-    parent_dir.mkdir(parents=True, exist_ok=True)
-  blob.download_to_filename(local_file_path)
-  return local_file_path
-
-
-def get_or_download_file(bucket: str,
-                         filename: str,
-                         use_anonymous_client: bool = False) -> str:
-  """Returns the local file path if the file already exists. 
-  Otherwise it downloads the file from GCS and returns the path it was downloaded to.
-  """
-  local_file_path = _get_local_path(bucket, filename)
-  if os.path.exists(local_file_path):
-    logging.info("Using already downloaded GCS file: %s", local_file_path)
-    return local_file_path
-  return download_file(bucket, filename, use_anonymous_client)
 
-
-def _get_local_path(bucketname: str, filename: str) -> str:
-  return os.path.join(TEMP_DIR, bucketname, filename)
+  bucket = storage_client.bucket(bucket_name)
+  blobs = bucket.list_blobs(prefix=blob_name)
+  count = 0
+  for blob in blobs:
+    if blob.name.endswith("/"):
+      continue
+    # Get the relative path to the input blob. This is used to download folder.
+    relative_path = os.path.relpath(blob.name, blob_name)
+    if relative_path == ".":
+      # In this case, the blob is a file.
+      local_file_path = local_path
+    else:
+      # In this case, the blob is a folder.
+      local_file_path = os.path.join(local_path, relative_path)
+    # Create the directory if it doesn't exist.
+    local_dir = os.path.dirname(local_file_path)
+    if not os.path.exists(local_dir):
+      os.makedirs(local_dir)
+    # Download the file.
+    blob.download_to_filename(local_file_path)
+    count += 1
+  if count == 0:
+    logging.warning("No object found from %s/%s", bucket_name, blob_name)
+    return False
+  return True
+
+
+def download_blob_by_path(gcs_path: str,
+                          local_path: str,
+                          use_anonymous_client: bool = False) -> bool:
+  """Downloads file/folder given full GCS path (i.e. gs://bucket/path/to/file)
+  to a local path.
+
+  Args:
+    gcs_path: The full GCS path (i.e. gs://bucket/path/to/file/).
+    local_path: The local path to download the blob to.
+  """
+  if not is_gcs_path(gcs_path):
+    raise ValueError(f"Invalid GCS path: {gcs_path}")
+  bucket_name, blob_name = get_path_parts(gcs_path)
+  return download_blob(bucket_name, blob_name, local_path, use_anonymous_client)
+
+
+def maybe_download(gcs_path: str,
+                   local_path_prefix='/tmp',
```

Comment:
nit: Any reason why `local_path_prefix` here and `local_path` in the other methods? Perhaps being more explicit and calling it something like `local_parent_dir` may be clearer.

Reply:
trying to be consistent with the naming, so local_path(*) means the download destination. I guess `local_path_root` maybe better here?

Reply:
That works.

Reply:
Fixed